pageserver/src/tenant.rs

```diff
@@ -2812,7 +2812,50 @@ impl Tenant {
         cancel: &CancellationToken,
         ctx: &RequestContext,
     ) -> anyhow::Result<Vec<Arc<Timeline>>> {
-        // grab mutex to prevent new timelines from being created here.
+        let timelines = self
+            .timelines
+            .lock()
+            .unwrap()
+            .values()
+            .filter(|tl| match target_timeline_id.as_ref() {
+                Some(required) => &tl.timeline_id == required,
+                None => true,
+            })
+            .cloned()
+            .collect::<Vec<_>>();
+
+        let mut gc_cutoffs: HashMap<TimelineId, GcCutoffs> =
+            HashMap::with_capacity(timelines.len());
+
+        // before taking the gc_cs lock, do the heavier weight finding of gc_cutoff points for
+        // currently visible timelines.
+        for timeline in timelines.iter() {
+            let cutoff = timeline
+                .get_last_record_lsn()
+                .checked_sub(horizon)
+                .unwrap_or(Lsn(0));
+
+            // FIXME: using questionmark here is wrong: either the whole tenant is shutting down,
+            // or a single timeline is getting deleted, we should at most log the error
+            let res = timeline.find_gc_cutoffs(cutoff, pitr, cancel, ctx).await;
+
+            match res {
+                Ok(cutoffs) => {
+                    let old = gc_cutoffs.insert(timeline.timeline_id, cutoffs);
+                    assert!(old.is_none());
+                }
+                Err(e) => {
+                    tracing::info!(timeline_id = %timeline.timeline_id, "ignoring failure to find gc cutoffs: {e:#}");
```

Comment:
This needs to be `warn!`: if find_gc_cutoffs is failing for some reason, we want to know at warn! level.

If you're afraid of tenant / timeline shutdowns being logged at the wrong level, we need distinguished erro rhandling.
But, `find_gc_cutoffs` returns `anyhow::Result` right now.

The solution, in my opinion, is not to try to verschlimmbessern it in this PR, but instead continue to use the questionmark as it was before this PR, and follow up with a refactor PR that moves the return type of `find_gc_cutoffs` to an enum error type with two variants (Cancelled, Other), at which point we can bring back this distinguished error handling here.

(Feel free to use this review comment as "reference in new issue")

Reply:
We don't need to log it at warn!:
- it's normal that timeline gets deleted during synthetic size calculation or gc
- it's also ok if a single timeline would start to fail for some other reason, like "key not found"

Neither of these will result in a failure which should stop gc or synthetic size calculation.

From the point of view of gc both will be safe, because we will:
- update the branchpoints on gc info
- not advance the cutoffs

Reply:
Well, I guess it could be a warn! would work, it'll be flaky with the tests so I'll need to test a few rounds, so I am not exactly enthusiastic about this.

Reply:
Aah yes, we don't enable metrics collection so perhaps only the gc could cause these.